noir/noir-repo/aztec_macros/src/transforms/note_interface.rs

```diff
@@ -0,0 +1,510 @@
+        //     ...
+        // }
+        //
+        // It assumes each field occupies an entire field and its serialized in definition order
+
+        if !note_impl.methods.iter().any(|(func, _)| func.def.name.0.contents == "fields") {
+            let note_fields_struct =
+                generate_note_fields_struct(&note_type, &note_fields, note_interface_impl_span)?;
+            note_fields_structs.push(note_fields_struct);
+            let note_fields_fn =
+                generate_note_fields_fn(&note_type, &note_fields, note_interface_impl_span)?;
+            note_impl.methods.push((note_fields_fn, note_impl.type_span));
+        }
+
+        // Automatically generate the header getter and setter methods
+        if !check_trait_method_implemented(trait_impl, "get_header") {
+            let get_header_fn = generate_note_get_header(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(get_header_fn));
+        }
+        if !check_trait_method_implemented(trait_impl, "set_header") {
+            let set_header_fn = generate_note_set_header(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(set_header_fn));
+        }
+
+        // Automatically generate the note type id getter method. The id itself its calculated as the concatenation
+        // of the conversion of the characters in the note's struct name to unsigned integers.
+        if !check_trait_method_implemented(trait_impl, "get_note_type_id") {
+            let get_note_type_id_fn =
+                generate_note_get_type_id(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(get_note_type_id_fn));
+        }
+
+        // Automatically generate the method to compute the note's content hash as:
+        // fn compute_note_content_hash(self: NoteType) -> Field {
+        //    // TODO(#1205) Should use a non-zero generator index.
+        //    dep::aztec::hash::pedersen_hash(self.serialize_content(), 0)
+        // }
+        //
+        if !check_trait_method_implemented(trait_impl, "compute_note_content_hash") {
+            let get_header_fn =
+                generate_compute_note_content_hash(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(get_header_fn));
+        }
+    }
+
+    module.types.extend(note_fields_structs);
+    Ok(())
+}
+
+fn generate_note_get_header(
+    note_type: &String,
+    impl_span: Option<Span>,
+) -> Result<NoirFunction, AztecMacroError> {
+    let function_source = format!(
+        "
+        fn get_header(note: {}) -> dep::aztec::note::note_header::NoteHeader {{
+            note.header
+        }}
+    ",
+        note_type
+    )
+    .to_string();
+
+    let (function_ast, errors) = parse_program(&function_source);
+    if !errors.is_empty() {
+        return Err(AztecMacroError::CouldNotImplementNoteSerialization {
```

Comment:
The error type here seems wrong? Same in the other generation functions

Reply:
Yeah, bad name (I kept the initial one). Refactored to `CouldNotImplementNoteInterface`